### PR TITLE
[12.0][ADD] web_widget_dropdown_dynamic

### PR DESCRIPTION
**Pull request by alexey-pelykh**
Dynamic dropdown widget that supports resolving options from backend of:

 * `fields.Char`
 * `fields.Integer`
 * `fields.Selection`

### CODE REVIEW

File: web_widget_dropdown_dynamic/readme/DESCRIPTION.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Dynamic dropdown widget that supports resolving options from backend.
```
Comment (pedrobaeza) [inline web_widget_dropdown_dynamic/readme/DESCRIPTION.rst]: Which field types are allowed in the widget?

Comment (alexey-pelykh) [reply web_widget_dropdown_dynamic/readme/DESCRIPTION.rst]: I plan to support character fields, initially. Maybe integer would be easy to support

Comment (pedrobaeza) [reply web_widget_dropdown_dynamic/readme/DESCRIPTION.rst]: Include selection as well

Comment (alexey-pelykh) [reply web_widget_dropdown_dynamic/readme/DESCRIPTION.rst]: What is usecase of that?

Comment (pedrobaeza) [reply web_widget_dropdown_dynamic/readme/DESCRIPTION.rst]: Restrict selection values according a criteria. Real example, in VAT declaration in Spain, you have 5 possible values in the result:

- To be debited in bank account
- To enter directly
- To return
- To be held in fiscal account
- To be compensated

First 2 values are selectable when the declaration is positive, and the other 2 when the declaration is negative. We need 3 fields (or at least 2) now for reflecting this + `attrs` in fields.

**Comment by pedrobaeza** [comment]
@Tardo should be the technical reviewer, although he is on holidays, so it will be a bit until he does it.

**Comment by foutoucour** [comment]
@alexey-pelykh could you add screenshots to illustrate the use cases? The module seems interesting but might be hard to understand the scope at the first read.
Thanks

**Comment by alexey-pelykh** [comment]
@foutoucour not sure there's a good screenshot to express it. Main usecase is: there's a field or fields (e.g. `amount`) and selection dropdown options depend if field is positive or negative. Another example: there's an login/password to third-party service and profile selection that is fetched after login/password is entered.

**Review by tarteo — APPROVED** [review]
The code LGTM, it's very clean and clear.

It would be nice to add some demo data to make functional testing easier.

**Comment by alexey-pelykh** [comment]
@tarteo another module with `_demo` suffix?

**Comment by alexey-pelykh** [comment]
@pedrobaeza does this module fit in your usecase?

**Comment by pedrobaeza** [comment]
I think so, but no time now for switching current approach that I exposed to use this new module. When migrating to v13 involved modules, we will be able to try it.

File: web_widget_dropdown_dynamic/static/src/js/field_dynamic_dropdown.js [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+         * @returns {jQuery}
+         */
+        getFocusableElement: function () {
+            return this.$el.is('select') ? this.$el : $();
+        },
+        /**
+         * @override
+         */
+        isSet: function () {
+            return this.value !== false;
```
Comment (daramousk) [inline web_widget_dropdown_dynamic/static/src/js/field_dynamic_dropdown.js:45]: Is this override needed?

**Review by daramousk — COMMENTED** [review]
Works great for Chars and Integers. If I try to use this on an already existing Selection field, and I try to return values that do not already exist on the `selection=[]` definition of that field, then I will get a `ValueError`. You can try this with the product type field.

**Comment by alexey-pelykh** [comment]
@daramousk but the case you've described effectively abuses the Selection field, so you should get ValueError.

File: web_widget_dropdown_dynamic/static/src/js/field_dynamic_dropdown.js (hunk 2 above)
Comment (alexey-pelykh) [reply web_widget_dropdown_dynamic/static/src/js/field_dynamic_dropdown.js:45]: Yes, `AbstractField` has different implementation of `isSet`: https://github.com/odoo/odoo/blob/856c2e9008f1af7bc1327d9c2900db1e109ab0fa/addons/web/static/src/js/fields/abstract_field.js#L276

**Comment by tarteo** [comment]
> @tarteo another module with `_demo` suffix?

Not necessarily you could use a field of a model which is introduced in `base`.

**Comment by alexey-pelykh** [comment]
@pedrobaeza any luck trying this module out?

**Comment by pedrobaeza** [comment]
Sorry, no

**Review by artemsenko — APPROVED** [review]
Functional test via https://github.com/OCA/bank-statement-import/pull/212

**Comment by alexey-pelykh** [comment]
@tarteo @pedrobaeza I've added a demo, please review and merge at earliest convenience

File: web_widget_dropdown_dynamic/readme/DESCRIPTION.rst [hunk 3]
```diff
@@ -0,0 +1,5 @@
+Dynamic dropdown widget that supports resolving options from backend of:
+
+ * `fields.Char`
+ * `fields.Integer`
+ * `fields.Selection`
```
Comment (pedrobaeza) [inline web_widget_dropdown_dynamic/readme/DESCRIPTION.rst]: I doubt this one is supported when you save a value that is not on the original selection.

Comment (alexey-pelykh) [reply web_widget_dropdown_dynamic/readme/DESCRIPTION.rst]: You are correct, saving a value that does not exist in original `selection` will give an error, which is by design. This field can be used only to _filter_ selection, not to extend. For fully-dynamic fields, use `Char` or `Integer`

Comment (pedrobaeza) [reply web_widget_dropdown_dynamic/readme/DESCRIPTION.rst]: You can put a note about that.

File: web_widget_dropdown_dynamic_example/__init__.py [hunk 4]
```diff
@@ -0,0 +1,3 @@
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+
+from . import models
```
Comment (pedrobaeza) [inline web_widget_dropdown_dynamic_example/__init__.py]: Line at the end

Comment (alexey-pelykh) [reply web_widget_dropdown_dynamic_example/__init__.py]: :) Don't we have a linter for that?

Comment (pedrobaeza) [reply web_widget_dropdown_dynamic_example/__init__.py]: That's the weird thing... It's not detected right now

Comment (alexey-pelykh) [reply web_widget_dropdown_dynamic_example/__init__.py]: Yep, 100%, on other modules as well

**Comment by alexey-pelykh** [comment]
@pedrobaeza comments attended

**Pull request merged**